### PR TITLE
Add RandomGamma for biomedical array transform

### PR DESCRIPTION
**Pull request by Fivethousand5k**
Thanks for your contribution and we appreciate it a lot. The following instructions would make your pull request more healthy and more easily get feedback. If you do not understand some items, don't worry, just make the pull request and seek help from maintainers.

## Motivation

Add support for random gamma correction for biomedical files

## Modification

add a new transform function named **RandomGamma**

## Checklist

1. Pre-commit or other linting tools are used to fix the potential lint issues.
2. The modification is covered by complete unit tests. If not, please add more unit test to ensure the correctness.
3. If the modification has potential influence on downstream projects, this PR should be tested with downstream projects, like MMDet or MMDet3D.
4. The documentation has been modified accordingly, like docstring or example tutorials.

### CODE REVIEW

File: mmseg/datasets/transforms/transforms.py [hunk 1]
```diff
@@ -1226,3 +1226,74 @@ def __repr__(self):
...
+        super().__init__()
+        assert 0 <= prob and prob <= 1
+        assert isinstance(gamma_range, tuple)
+        self.prob = prob
+        self.gamma_range = gamma_range
+
+    @cache_randomness
+    def generate_gamma(self):
+        return np.random.rand() < self.prob, np.random.uniform(
+            min(*self.gamma_range), max(*self.gamma_range))
```
Comment (MeowZheng) [inline mmseg/datasets/transforms/transforms.py:1261]:
```suggestion
    @cache_randomness
    def _do_gamma(self):
        return np.random.rand() < self.prob

    @cache_randomness
    def _generate_gamma(self):
        return np.random.uniform(
            min(*self.gamma_range), max(*self.gamma_range))
```

File: mmseg/datasets/transforms/transforms.py [hunk 2]
```diff
@@ -1226,3 +1226,74 @@ def __repr__(self):
...
+        """Call function to randomly adjust the contrast of image.
+
+        Args:
+            results (dict): Result dict.
+
+        Returns:
+            dict: Result dict with random gamma transformed.
+        """
+        do_gamma, gamma = self.generate_gamma()
+        if do_gamma:
```
Comment (MeowZheng) [inline mmseg/datasets/transforms/transforms.py:1273]:
```suggestion
        do_gamma = self._do_gamma()
        if do_gamma:
            gamma = self._generate_gamma()
```

File: mmseg/datasets/transforms/transforms.py [hunk 3]
```diff
@@ -1226,3 +1226,74 @@ def __repr__(self):
         repr_str += f'edge_width={self.edge_width}, '
         repr_str += f'ignore_index={self.ignore_index})'
         return repr_str
+
+
+@TRANSFORMS.register_module()
+class RandomGamma(BaseTransform):
+    """Using random gamma correction to process the image.
+
+    Modified from https://docs.monai.io/en/0.3.0/_modules/monai/transforms/intensity/array.html#RandAdjustContrast
```
Comment (MeowZheng) [inline mmseg/datasets/transforms/transforms.py:1235]: please add copyright and license info about your reference code

File: mmseg/datasets/transforms/transforms.py [hunk 4]
```diff
@@ -1226,3 +1226,74 @@ def __repr__(self):
...
+        """
+        do_gamma, gamma = self.generate_gamma()
+        if do_gamma:
+            results['img'] = self._adjust_contrast(results['img'], gamma)
+        return results
+
+    def _adjust_contrast(self, img: np.ndarray, gamma: float):
+        """Adjust contrast according to the gamma:
+
+        x =  ((x - min) / intensity_range) ^ gamma * intensity_range + min
```
Comment (MeowZheng) [inline mmseg/datasets/transforms/transforms.py:1280]: please follow reStructured

File: mmseg/datasets/transforms/transforms.py [hunk 5]
```diff
@@ -1226,3 +1226,74 @@ def __repr__(self):
...
+
+    - img
+
+    Args:
+        prob (float): RandomGamma probability.
+        gamma_range: Range of gamma values.
+    """
+
+    def __init__(self, prob: float,
+                 gamma_range: tuple[float] = (0.5, 4.5)) -> None:
```
Comment (MeowZheng) [inline mmseg/datasets/transforms/transforms.py:1251]: Does it should handle input gamma as `int` type

File: mmseg/datasets/transforms/transforms.py [hunk 6]
```diff
@@ -1226,3 +1226,74 @@ def __repr__(self):
...
+        super().__init__()
+        assert 0 <= prob and prob <= 1
+        assert isinstance(gamma_range, tuple)
```
Comment (MeowZheng) [inline mmseg/datasets/transforms/transforms.py:1254]: do we need to add a requirement about gamma>0.5?

File: mmseg/datasets/transforms/transforms.py [hunk 7]
```diff
@@ -1226,3 +1226,74 @@ def __repr__(self):
         repr_str += f'edge_width={self.edge_width}, '
         repr_str += f'ignore_index={self.ignore_index})'
         return repr_str
+
+
+@TRANSFORMS.register_module()
+class RandomGamma(BaseTransform):
+    """Using random gamma correction to process the image.
```
Comment (MeowZheng) [inline mmseg/datasets/transforms/transforms.py:1233]:
```suggestion
    """Using random gamma correction to process the biomedical image.
```

File: mmseg/datasets/transforms/transforms.py [hunk 8]
```diff
@@ -1226,3 +1226,74 @@ def __repr__(self):
...
+
+    Modified from https://docs.monai.io/en/0.3.0/_modules/monai/transforms/intensity/array.html#RandAdjustContrast
+
+    Required Keys:
+    
+    - img
```
Comment (MeowZheng) [inline mmseg/datasets/transforms/transforms.py:1239]:
```suggestion
    - img (np.ndarray, float)
```

File: mmseg/datasets/transforms/transforms.py [hunk 9]
```diff
@@ -1226,3 +1226,74 @@ def __repr__(self):
...
+
+    Modified Keys:
+
+    - img
+
+    Args:
+        prob (float): RandomGamma probability.
+        gamma_range: Range of gamma values.
```
Comment (MeowZheng) [inline mmseg/datasets/transforms/transforms.py:1247]: please add type for `gamma_range`

**Review by MeowZheng — COMMENTED** [review]
please fix lint problem, and add unit test

File: mmseg/datasets/transforms/transforms.py (hunk 6 above)
Comment (Fivethousand5k) [reply mmseg/datasets/transforms/transforms.py]: Yeah, I have added such an assertion in the latest commit.

File: mmseg/datasets/transforms/transforms.py [hunk 10]
```diff
@@ -1226,3 +1226,86 @@ def __repr__(self):
...
+
+    - img
+
+    Args:
+        prob (float): RandomGamma probability.
+        gamma_range (tuple[float]): Range of gamma values.
+    """
+
+    def __init__(self, prob: float,
+                 gamma_range: tuple[float] = (0.5, 4.5)) -> None:
```
Comment (MeowZheng) [inline mmseg/datasets/transforms/transforms.py:1255]:
```suggestion
from typing import Sequence
                 gamma_range: Sequence[float] = (0.5, 4.5)) -> None:
```

File: mmseg/datasets/transforms/transforms.py [hunk 11]
```diff
@@ -1226,3 +1226,86 @@ def __repr__(self):
...
+        Args:
+            results (dict): Result dict.
+
+        Returns:
+            dict: Result dict with random gamma transformed.
+        """
+        do_gamma = self._do_gamma
+        if do_gamma:
+            gamma = self._generate_gamma()
+            results['img'] = self._adjust_contrast(results['img'], gamma)
```
Comment (MeowZheng) [inline mmseg/datasets/transforms/transforms.py:1284]:
```suggestion
        if do_gamma:
            self.gamma = self._generate_gamma()
            results['img'] = self._adjust_contrast(results['img'], self.gamma)
        else:
            self.gamma = None
```

File: mmseg/datasets/transforms/transforms.py [hunk 12]
```diff
@@ -1226,3 +1226,86 @@ def __repr__(self):
...
+        epsilon = 1e-7
+        img_min = img.min()
+        img_range = img.max() - img_min
+        return np.power(((img - img_min) / float(img_range + epsilon)),
+                        gamma) * img_range + img_min
+
+    def __repr__(self):
+        repr_str = self.__class__.__name__
+        repr_str += f'(prob={self.prob}, '
+        repr_str += f'gamma_range={self.gamma_range})'
```
Comment (MeowZheng) [inline mmseg/datasets/transforms/transforms.py:1310]:
```suggestion
        repr_str += f'gamma_range={self.gamma_range}, '
        repr_str += f'gamma={self.gamma})'
```

File: mmseg/datasets/transforms/transforms.py [hunk 13]
```diff
@@ -1226,3 +1226,86 @@ def __repr__(self):
...
+
+        Returns:
+            np.ndarray: Image after contrast change.
+        """
+
+        epsilon = 1e-7
+        img_min = img.min()
+        img_range = img.max() - img_min
+        return np.power(((img - img_min) / float(img_range + epsilon)),
+                        gamma) * img_range + img_min
```
Comment (MeowZheng) [inline mmseg/datasets/transforms/transforms.py:1305]: if this `float()` convert necessary?

**Comment by Fivethousand5k** [comment]
I will raise another pull request for this transform which follows the design of the nnUNet. 
Thank you for your time.

**Pull request closed without merging**